File: src/main/java/org/frawa/elmtest/core/json/Problem.java
package org.frawa.elmtest.core.json;

import com.google.gson.JsonElement;

import java.util.List;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Problem {
    public String title;
    public Region region;
    public List<JsonElement> message;

    public String getTextMessage() {
        Predicate<JsonElement> hasText = element ->
                element.isJsonPrimitive()
                        || (element.isJsonObject() && element.getAsJsonObject().has("string"));

        Function<JsonElement, String> toText = element ->
                element.isJsonPrimitive()
                        ? element.getAsString()
                        : element.getAsJsonObject().get("string").getAsString();

        return message.stream()
                .filter(hasText)
                .map(toText)
                .collect(Collectors.joining());
    }


}

File: src/main/kotlin/org/elm/ide/test/core/json/Problem.kt
package org.elm.ide.test.core.json

import com.google.gson.JsonElement

class Problem {
    var title: String? = null
    var region: Region? = null
    var message: List<JsonElement>? = null

    val textMessage: String
        get() {
            val hasText = { element:JsonElement -> element.isJsonPrimitive || element.isJsonObject && element.asJsonObject.has("string") }

            val toText = { element:JsonElement ->
                if (element.isJsonPrimitive)
                    element.asString
                else
                    element.asJsonObject.get("string").asString
            }

            return message!!
                    .filter(hasText).joinToString("", transform = toText)
        }


}
